### DISCUSSION TITLE
Legend ordering

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [observablehq/plot#1953] [Q&A] [answered]
When showing a legend, the names are ordered alphabetically. How do I make a custom order?

I can't find this info in the docs or anywhere else. I found something about setting `domain` to order facets, but I can't get it working for the legend.

The plotting code is essentially:

    Plot.plot({
        color: {
          legend: true,
        },

        marks: [
            Plot.lineY(data, {x: "Date", y: "Value", stroke: "Name"}),
        ],
    });

Thanks!

**Comment by Fil** [comment]
Please share a notebook demonstrating the issue. Thanks!

**Comment by mbostock** [comment, marked as answer]
Per [Scales](https://observablehq.com/plot/features/scales#type-inference):

> For an ordinal scale, the domain defaults to the sorted union (all distinct values in natural order) of associated values; see the [sort mark option](https://observablehq.com/plot/features/scales#sort-mark-option) to change the order.

The _color_ scale here is ordinal (assuming that the _Name_ column is full of string names), so that’s why you’re getting alphabetical order for the legend. If you want a custom order, you can either use the **sort** option on the lineY mark, or you can set the **domain** option on the _color_ scale definition. E.g.,

```js
Plot.plot({
  color: {
    legend: true,
    domain: ["small", "medium", "large"]
  },
  marks: [
    Plot.lineY(data, {x: "Date", y: "Value", stroke: "Name"}),
  ]
});
```

As @Fil said, if you want help, it’s generally best to share a notebook with a minimal reproduction of the problem you’re having. That way we can see exactly the behavior and are more likely to diagnose the problem.

**Reply by Hvass-Labs** [reply]
Thanks very much! I actually tried that, but I think I had accidentally set `domain: null` in some cases as I make nearly 20 plots for one web-page, and that gave a strange error-message. But it works now when I do it properly.

However, I only knew about setting `domain` because you had mentioned it deep inside another discussion on facets. So please consider adding this to the docs.

I can see it would be much easier for you to diagnose my problem if I make a Notebook on HQ. But my plotting code is entangled deep inside a web-application and the data comes from a database, so it's quite a hassle to turn that into a Notebook. But maybe I could find some of your code-samples and modify them to show my problems in the future.

Merry Christmas!

**Reply by mbostock** [reply]
> So please consider adding this to the docs.

For what it’s worth, I quoted and linked the [relevant part of the docs](https://observablehq.com/plot/features/scales#type-inference) above.

**Reply by Hvass-Labs** [reply]
Yes, I noticed you linked to the docs for the `sort` option, but the `domain` option for ordering the legend doesn't seem to be documented? Perhaps I'm wrong, but as a user searching for this, I would expect to find it in the [docs for legends](https://observablehq.com/plot/features/legends). The docs for scales is very long so I just browse and search it, but it doesn't contain the word "legend" at all.

With that said, when I forget something in e.g. Python or some code-library, I usually just find some code-sample on StackOverflow to remind me, instead of reading the official docs. So these code snippets in GitHub issues and discussions that you write are also helpful in addition to your official docs. Your time is not wasted writing these.

**Reply by mbostock** [reply]
I see. I think it’s a conceptual issue: this isn’t a legend option because the legend is simply reflecting the definition of the underlying scale. You’re thinking of this as “ordering the legend” but we don’t support that — the legend merely truthfully reflects the scale, and so the only way to do this is to change how the scale is defined. We can add a note from legends that points people back to scales.

**Reply by Hvass-Labs** [reply]
OK, I see.

I generally get the impression that "scales" are central to a lot of what you do, so it would be wise of users to understand it properly. But that particular doc-page is just so long and abstract that it can be a bit daunting if a user just wants to quickly do something simple like reordering the text in a legend. Most users probably don't want or need to have a deep understanding of how your sophisticated API works. They just want to copy-paste-modify code to make their plot and get on with their work.

You said in another thread a few years ago, that it is quite challenging to write docs that everyone likes. I agree and it is especially true when you are an expert in your field, because it is hard to see what a beginner finds difficult to understand.

Maybe the solution is to just supplement your docs with code-samples in these discussions whenever people are confused, and accept that people sometimes ask you "stupid" and annoying questions :-)

Eventually you will have your own little StackOverflow here with most questions answered. It is important to give near-complete code-samples as you've done here, and not just vague descriptions of how to solve a problem, because that is also very confusing to a beginner - just as it is confusing when the user gives you a vague description of their problem, instead of supplying you with a Notebook :-)
